resolwe/flow/models/fields.py

```diff
@@ -0,0 +1,158 @@
+            # ResolweSlugField(populate_from='foo')
+            attr = getattr(instance, self.populate_from)
+            return attr() if callable(attr) else attr
+
+    def pre_save(self, instance, add):
+        """Ensure slug uniqunes before save."""
+        slug = self.value_from_object(instance)
+
+        if not slug and self.populate_from:
+            slug = self._get_populate_from_value(instance)
+
+        if slug:
+            slug = slugify(slug)
+
+        if not slug:
+            slug = None
+
+            if not self.blank:
+                slug = instance._meta.model_name  # pylint: disable=protected-access
+            elif not self.null:
+                slug = ''
+
+        if slug:
+            # Make sure that potential sequence added won't excede maximal length.
+            slug = slug[:(self.max_length - 10)]
+
+            constraints_placeholder, constraints_values = self._get_unique_constraints(instance)
+
+            instance_pk_name = instance._meta.pk.name  # pylint: disable=protected-access
+
+            query_params = {
+                'constraints_placeholder': constraints_placeholder,
+                'slug_column': self.column,
+                'table_name': self.model._meta.db_table,  # pylint: disable=protected-access
+                'pk_neq_placeholder': 'AND {} != %s'.format(instance_pk_name) if instance.pk else ''
+            }
+
+            query_escape_params = [slug]
+            if instance.pk:
+                query_escape_params.append(instance.pk)
+            query_escape_params.extend(constraints_values)
+            query_escape_params.append(slug)
+            query_escape_params.append('^{}(-[0-9]*)?$'.format(slug))
+            if instance.pk:
+                query_escape_params.append(instance.pk)
+            query_escape_params.extend(constraints_values)
+
+            with connection.cursor() as cursor:
+                # TODO: Slowest part of this query is `MAX` function. It can
+                #       be optimized by indexing slug column by slug sequence.
+                #       https://www.postgresql.org/docs/9.4/static/indexes-expressional.html
+                cursor.execute(
+                    """
+                    SELECT
+                        CASE
+                            WHEN (
+                                EXISTS(
+                                    SELECT 1 FROM {table_name} WHERE (
+                                        {slug_column} = %s
+                                        {pk_neq_placeholder}
+                                        {constraints_placeholder}
+                                    )
+                                )
+                            ) THEN MAX(slug_sequence) + 1
+                            ELSE NULL
+                        END
+                    FROM (
+                        SELECT COALESCE(
+                            NULLIF(
+                                RIGHT({slug_column}, -length(%s)-1),
```

Comment:
Isn't this length constant? Any reason why you don't compute it in Python and just provide the actual value here?